stagecoach/price-pusher/price_pusher/core/listener.py

```diff
@@ -0,0 +1,244 @@
+        price_config: PriceConfig,
+        polling_frequency_in_s: DurationInSeconds,
+    ) -> None:
+        self.id = generate_human_readable_id()
+        self.request_handler = request_handler
+        self.price_config = price_config
+
+        self.oracle_prices = {}
+        self.orchestrator_prices = None
+
+        self.notification_event = asyncio.Event()
+        self.polling_frequency_in_s = polling_frequency_in_s
+
+        self._log_listener_spawning()
+
+    async def run_forever(self) -> None:
+        """
+        Main loop responsible of:
+            - fetching the latest oracle prices
+            - checking if the oracle needs update
+            - pushing notification to the orchestration if it does.
+        """
+        last_fetch_time = -1
+        while True:
+            current_time = asyncio.get_event_loop().time()
+            if current_time - last_fetch_time >= self.polling_frequency_in_s:
+                await self._fetch_all_oracle_prices()
+                last_fetch_time = current_time
+            if await self._does_oracle_needs_update():
+                self._notify()
+                last_fetch_time = -1
+            # Check every second if the oracle needs an update
+            await asyncio.sleep(1)
+
+    def set_orchestrator_prices(self, orchestrator_prices: dict) -> None:
+        """
+        Set the reference of the orchestrator prices in the Listener.
+        """
+        self.orchestrator_prices = orchestrator_prices
+
+    async def _fetch_all_oracle_prices(self) -> None:
+        """
+        Fetch the latest oracle prices for all assets in parallel.
+        """
+        tasks = [
+            self.request_handler.fetch_latest_asset_price(asset)
+            for asset in self.price_config.get_all_assets()
+        ]
+        results = await asyncio.gather(*tasks, return_exceptions=True)
+
+        for entry in results:
+            if isinstance(entry, Exception) or isinstance(entry, PragmaAPIError):
+                logger.error(f"Error fetching oracle price: {entry}")
+                continue
+            if entry is None:
+                continue
+
+            pair_id = entry.get_pair_id().replace(",", "/")
+            asset_type = "SPOT" if isinstance(entry, SpotEntry) else "FUTURE"
+
+            if pair_id not in self.oracle_prices:
+                self.oracle_prices[pair_id] = {}
+
+            self.oracle_prices[pair_id][asset_type] = entry
+
+    def _get_most_recent_orchestrator_entry(
+        self, pair_id: str, asset_type: AssetType
+    ) -> Optional[Entry]:
+        """
+        Retrieves the latest registered entry from the orchestrator prices.
+        """
+        if self.orchestrator_prices is None:
+            raise ValueError("Orchestrator must set the prices dictionnary.")
+        entries = [entry for entry in self.orchestrator_prices[pair_id][asset_type].values()]
+        return max(entries, key=lambda entry: entry.base.timestamp, default=None)
```

Comment:
there is an edge case where this will throw an error because `self.orchestrator_prices[pair_id] = {}`

Reply:
just double check that it's handled correctly

Reply:
done